bundles/org.openhab.core/src/main/java/org/openhab/core/internal/items/MetadataStateDescriptionFragmentProvider.java

```diff
@@ -0,0 +1,151 @@
+
+    private final Logger logger = LoggerFactory.getLogger(MetadataStateDescriptionFragmentProvider.class);
+
+    public static final String STATEDESCRIPTION_METADATA_NAMESPACE = "stateDescription";
+
+    private final MetadataRegistry metadataRegistry;
+
+    private final Integer rank;
+
+    @Activate
+    public MetadataStateDescriptionFragmentProvider(final @Reference MetadataRegistry metadataRegistry,
+            Map<String, Object> properties) {
+        this.metadataRegistry = metadataRegistry;
+
+        Object serviceRanking = properties.get(Constants.SERVICE_RANKING);
+        if (serviceRanking instanceof Integer) {
+            rank = (Integer) serviceRanking;
+        } else {
+            rank = 1; // takes precedence over other providers usually ranked 0
+        }
+    }
+
+    @Override
+    public @Nullable StateDescriptionFragment getStateDescriptionFragment(@NonNull String itemName,
+            @Nullable Locale locale) {
+        Metadata metadata = metadataRegistry.get(new MetadataKey(STATEDESCRIPTION_METADATA_NAMESPACE, itemName));
+
+        if (metadata != null) {
+            try {
+                StateDescriptionFragmentBuilder builder = StateDescriptionFragmentBuilder.create();
+                if (metadata.getConfiguration().containsKey("pattern")) {
+                    builder.withPattern((String) metadata.getConfiguration().get("pattern"));
+                }
+                if (metadata.getConfiguration().containsKey("min")) {
+                    builder.withMinimum(getBigDecimal(metadata.getConfiguration().get("min")));
+                }
+                if (metadata.getConfiguration().containsKey("max")) {
+                    builder.withMaximum(getBigDecimal(metadata.getConfiguration().get("min")));
+                }
+                if (metadata.getConfiguration().containsKey("step")) {
+                    builder.withStep(getBigDecimal(metadata.getConfiguration().get("min")));
+                }
+                if (metadata.getConfiguration().containsKey("readOnly")) {
+                    builder.withReadOnly(getBoolean(metadata.getConfiguration().get("readOnly")));
+                }
+                if (metadata.getConfiguration().containsKey("options")) {
+                    List<StateOption> stateOptions = Stream
+                            .of(metadata.getConfiguration().get("options").toString().split(",")).map(o -> {
+                                return (o.contains("="))
+                                        ? new StateOption(o.split("=")[0].trim(), o.split("=")[1].trim())
+                                        : new StateOption(o.trim(), null);
+                            }).collect(Collectors.toList());
+                    builder.withOptions(stateOptions);
+                }
+
+                return builder.build();
+            } catch (Exception e) {
+                logger.error("Unable to parse the stateDescription from metadata for item {}", itemName, e);
```

Comment:
This should imho be a WARN for the user, telling him that the value is ignored. No stacktrace must be added here as we are not talking about a bug in the software.